klippy/extras/bed_mesh.py

```diff
@@ -573,6 +573,112 @@ def _verify_algorithm(self, error):
+                  self.orig_config["mesh_min"][1]))
+
+        gcmd.respond_info("Original mesh bounds: (%s,%s)" %
+                          (self.orig_config["mesh_min"],
+                           self.orig_config["mesh_max"]))
+        gcmd.respond_info("Original probe count: (%s,%s)" %
+                          (self.mesh_config["x_count"],
+                           self.mesh_config["y_count"]))
+        gcmd.respond_info("Adapted mesh bounds: (%s,%s)" %
+                          (adjusted_mesh_min, adjusted_mesh_max))
+        gcmd.respond_info("Ratio: (%s, %s)" % ratio)
+
+        new_x_probe_count = int(
+            math.ceil(self.mesh_config["x_count"] * ratio[0]))
+        new_y_probe_count = int(
+        math.ceil(self.mesh_config["y_count"] * ratio[1]))
+
+        # There is one case, where we may have to adjust the probe counts:
+        # axis0 < 4 and axis1 > 6 (see _verify_algorithm).
+        min_num_of_probes = 3
+        if max(new_x_probe_count, new_y_probe_count) > 6 and \
+           min(new_x_probe_count, new_y_probe_count) < 4:
+            min_num_of_probes = 4
+
+        new_x_probe_count = max(min_num_of_probes, new_x_probe_count)
+        new_y_probe_count = max(min_num_of_probes, new_y_probe_count)
+
+        gcmd.respond_info("Adapted probe count: (%s,%s)" %
+                          (new_x_probe_count, new_y_probe_count))
+
+        # If the adapted mesh size is too small, adjust it to something
+        # useful.
+        adjusted_mesh_size = (max(adjusted_mesh_size[0], new_x_probe_count),
+                              max(adjusted_mesh_size[1], new_y_probe_count))
+
+        if self.radius is not None:
+            adapted_radius = math.sqrt((adjusted_mesh_size[0] ** 2) +
+                                       (adjusted_mesh_size[1] ** 2)) / 2
+            adapted_origin = (adjusted_mesh_min[0] +
+                              (adjusted_mesh_size[0] / 2),
+                              adjusted_mesh_min[1] +
+                              (adjusted_mesh_size[1] / 2))
+            to_adapted_origin = math.sqrt(adapted_origin[0]**2 +
+                                          adapted_origin[1]**2)
+            # If the adapted mesh size is smaller than the default/full
+            # mesh, adjust the parameters. Otherwise, just do the full mesh.
+            if adapted_radius + to_adapted_origin < self.radius:
+                self.radius = adapted_radius
+                self.origin = adapted_origin
+                self.mesh_min = (-self.radius, -self.radius)
+                self.mesh_max = (self.radius, self.radius)
+                self.mesh_config["x_count"] = self.mesh_config["y_count"] = \
+                        max(new_x_probe_count, new_y_probe_count)
+        else:
+            self.mesh_min = adjusted_mesh_min
+            self.mesh_max = adjusted_mesh_max
+            self.mesh_config["x_count"] = new_x_probe_count
+            self.mesh_config["y_count"] = new_y_probe_count
```

Comment:
```suggestion
            self.mesh_config["y_count"] = new_y_probe_count
        self._profile_name = None
```

Reply:
To disable the mesh autosave we can set `self._profile name` to `None` here, then we can change line 892:
```python
if self._profile_name is not None:
    self.bedmesh.save_profile(self._profile_name)
```

Reply:
> To disable the mesh autosave we can set `self._profile name` to `None` here, then we can change line 892:
> 
> ```python
> if self._profile_name is not None:
>     self.bedmesh.save_profile(self._profile_name)
> ```

Wouldn't calling `BED_MESH_PROFILE SAVE=...` still end up saving the adapted mesh? I think we would have to prevent this from happening, as well.

Reply:
> To disable the mesh autosave we can set `self._profile name` to `None` here, then we can change line 892:
> 
> ```python
> if self._profile_name is not None:
>     self.bedmesh.save_profile(self._profile_name)
> ```

What if we set a flag in `set_adaptive_mesh()` (`self.is_adaptive = True`) and then in `ProfileManager.save_config()` we check if it's True. If so, display an error message?

Reply:
This is the proposed diff that should take care of both the auto- and explicit save:

```diff
diff --git a/klippy/extras/bed_mesh.py b/klippy/extras/bed_mesh.py
index bc053e27..2853ed46 100644
--- a/klippy/extras/bed_mesh.py
+++ b/klippy/extras/bed_mesh.py
@@ -90,6 +90,7 @@ class BedMesh:
         self.printer.register_event_handler("klippy:connect",
                                             self.handle_connect)
         self.last_position = [0., 0., 0., 0.]
+        self._is_adaptive = False
         self.bmc = BedMeshCalibrate(config, self)
         self.z_mesh = None
         self.toolhead = None
@@ -216,6 +217,8 @@ class BedMesh:
                     raise self.gcode.error(
                         "Mesh Leveling: Error splitting move ")
         self.last_position[:] = newpos
+    def set_adaptive(self, adaptive):
+        self._is_adaptive = adaptive
     def get_status(self, eventtime=None):
         return self.status
     def update_status(self):
@@ -225,6 +228,7 @@ class BedMesh:
             "mesh_max": (0., 0.),
             "probed_matrix": [[]],
             "mesh_matrix": [[]],
+            "adaptive" : self._is_adaptive,
             "profiles": self.pmgr.get_profiles()
         }
         if self.z_mesh is not None:
@@ -678,6 +682,7 @@ class BedMeshCalibrate:
             self.mesh_max = adjusted_mesh_max
             self.mesh_config["x_count"] = new_x_probe_count
             self.mesh_config["y_count"] = new_y_probe_count
+        self.bedmesh.set_adaptive(True)
         return True
     def update_config(self, gcmd):
         # reset default configuration
@@ -690,6 +695,7 @@ class BedMeshCalibrate:
 
         params = gcmd.get_command_parameters()
         need_cfg_update = False
+        self.bedmesh.set_adaptive(False)
         if self.radius is not None:
             if "MESH_RADIUS" in params:
                 self.radius = gcmd.get_float("MESH_RADIUS")
@@ -889,7 +895,8 @@ class BedMeshCalibrate:
             z_mesh.set_zero_reference(*self.zero_ref_pos)
         self.bedmesh.set_mesh(z_mesh)
         self.gcode.respond_info("Mesh Bed Leveling Complete")
-        self.bedmesh.save_profile(self._profile_name)
+        if not self.bedmesh.get_status().get("adaptive", False):
+            self.bedmesh.save_profile(self._profile_name)
     def _dump_points(self, probed_pts, corrected_pts, offsets):
         # logs generated points with offset applied, points received
         # from the finalize callback, and the list of corrected points
@@ -1332,6 +1339,9 @@ class ProfileManager:
                 "file and restart the printer" %
                 (('\n').join(self.incompatible_profiles)))
     def save_profile(self, prof_name):
+        if self.bedmesh.get_status().get("adaptive", False):
+            self.gcode.respond_info("Adaptive meshes cannot be saved.")
+            return
         z_mesh = self.bedmesh.get_mesh()
         if z_mesh is None:
             self.gcode.respond_info(
```

Reply:
I don't think its necessary to prevent users from explicitly saving the profile.  That is completely opt-in, and perhaps they may have some good reason to do it.  My guess is that for the general user it won't be an issue, because adaptive meshing will only occur during a print.  If they are using a supplied configuration then `BED_MESH_CALIBRATE ADAPTIVE=1` will be in the start macro and `BED_MESH_CLEAR` will be in the cancel and end macros, reducing the chance of unintentionally saving.

Reply:
> I don't think its necessary to prevent users from explicitly saving the profile. That is completely opt-in, and perhaps they may have some good reason to do it. My guess is that for the general user it won't be an issue, because adaptive meshing will only occur during a print. If they are using a supplied configuration then `BED_MESH_CALIBRATE ADAPTIVE=1` will be in the start macro and `BED_MESH_CLEAR` will be in the cancel and end macros, reducing the chance of unintentionally saving.

OK. I use your suggested change then.

Reply:
@Arksine, done.